### ISSUE TITLE
Understanding inner optimizer parameters through MAML example

### ISSUE THREAD

**Issue by kgarg8** [facebookresearch/higher#74]
Hi,

I was going through MAML example in the repo. One thing that struck me was that we are using **net.parameters()** in the inner optimizer through this line (which I believe represent **meta-parameters** for the model):
https://github.com/facebookresearch/higher/blob/5a4dee462c670b25f999505d7d6a32c4cc873fc7/examples/maml-omniglot.py#L130

Won't diffopt.step update the meta_parameters also through the above line?

**Comment by egrefen** [comment]
Hello @kgarg8. Thanks for your question. I understand you are asking why the inner optimizer defined here:
https://github.com/facebookresearch/higher/blob/5a4dee462c670b25f999505d7d6a32c4cc873fc7/examples/maml-omniglot.py#L130
has the same sort of definition as the meta optimizer defined here:
https://github.com/facebookresearch/higher/blob/5a4dee462c670b25f999505d7d6a32c4cc873fc7/examples/maml-omniglot.py#L103

The `inner_opt` is a normal pytorch optimizer instance, and when we create it, we are declaring the sort of optimizer we would wish to use when unrolling the inner optimization loop for the purpose of MAML (but can't due to current limitations in pytorch). When you create the differentiable version of the model and optimizer for the purpose of this unroll on these lines
https://github.com/facebookresearch/higher/blob/5a4dee462c670b25f999505d7d6a32c4cc873fc7/examples/maml-omniglot.py#L135-L138
The library branches of the states of `net` and `inner_opt` to create functional copies, and under the hood starts tracking the "temporary" (unrolled) parameters of `fnet` within `diffopt`. When you take a meta-gradient step on this line:
https://github.com/facebookresearch/higher/blob/5a4dee462c670b25f999505d7d6a32c4cc873fc7/examples/maml-omniglot.py#L164
The meta-gradient backpropagates to the meta-variable, which in the case of MAML is the original weights of the model before the unrolled inner loop.

I hope this clarifies things, and will close the issue, but if you need more information and have follow up questions, please don't hesitate to re-open it!

**Comment by kgarg8** [comment]
Thanks, it makes sense to me now.
